### ISSUE TITLE
Should we settle on a single language for automation scripts? Which one?

### ISSUE THREAD

**Issue by larsks** [operate-first/apps#637]
From the discussion @HumairAK started in #635:

> one issue I'm seeing is right now we have a few scripts in bash, one in python (that @larsks has worked on) we will need to narrow down to one language if we are to wrap these, lest we start calling bash / python scripts from golang or something (yuck).

**Comment by larsks** [comment]
I don't feel strongly about settle on a single language (or not), but I would love to see us stop parsing YAML in bash, especially given the existence of two `yq` commands with mutually incompatible syntax.

**Comment by HumairAK** [comment]
I'm happy with golang, working as closely as we do with openshift/k8s it could be valuable experience. Also, we may want to extend the tooling to also communicate with OCP and thus client-go support could serve useful. But I don't know how suitable Go is for scripts, since it requires compilation, it may add additional overhead. But on the other hand, wrapping these go scripts into one cli/program in the future would become a lot easier. We could, for now, just settle for running these via `go run`.

**Comment by tumido** [comment]
Yeah, I'd like to bring us closer to clis like `kubectl`, `kustomize`, `argocd` etc. We can easily include those as subcommands if we want to since it's all golang libs.
Other tool we might want to consider including or extending is [kam](https://github.com/redhat-developer/kam).

I think the necessity to compile go binaries actually makes it more portable and it's useful. It's very easy to build go binary for various platforms without any overhead (it's just 2 build flags, that's it). It also means there's zero dependencies - no Python version constraint or pip dependencies required, proper windows support if we want it to, etc.

I think an immerse interactive CLI would be an appreciated end result.

**Comment by tumido** [comment]
Note: My previous comment was more focused on the "end goal" for our "immerse" CLI. Which is maybe further down the road than this issue... In the end I think we should have a well defined end goal (that I can fork from the previous comment) and all our refactoring should keep that in mind and do small steps towards that direction.

So.. if our end goal is an extendable interactive CLI, I'd say it would be for the best to use `golang` for that, hence refactor the scripts into the same language as the future CLI so it aligns with the end goal here.

**Comment by HumairAK** [comment]
Skimming through the `kam` tool...this seems extremely heavily opinionated of things that we are also very heavily opinionated on.

**Comment by tumido** [comment]
> this seems extremely heavily opinionated

Yeah, I've reached the same conclusion @HumairAK :+1:

**Comment by HumairAK** [comment]
further thinking on this, why not just do a single binary right off the bat and separate what would have been different scripts in to different sub commands, e.g: 

```
./cli [subcommand] ....<additional flags or sub-sub-commands> 
```
So like: 
```
./cli onboarding blah-blah
./cli enable-monitoring blah-blah
./cli update-cluster-version blah-blah
```

Then re-using common code will be a bit clearer/intuitive. I think folks will also stay more consistent in their formatting as they are likely to copy the standards set in the existing code.

**Comment by tumido** [comment]
Yeah, that sounds good :slightly_smiling_face:
